### PR TITLE
[SHELL32] Fix CShellLink::GetIconLocation

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
This will fix CORE-14961. It was regression by #784.
JIRA issue: [CORE-14961](https://jira.reactos.org/browse/CORE-14961), [CORE-14009](https://jira.reactos.org/browse/CORE-14009)

### CODE REVIEW

**Comment by katahiromz** [comment]
```c
if (FAILED(hr))
```
We can modify the failure condition.

**Comment by katahiromz** [comment]
BEFORE MOD:
before-mod
AFTER MOD:
after-mod

Let's go without further modification.

**Comment by HBelusca** [comment]
We just note the extra test failures.

**Comment by katahiromz** [comment]
Getting better...
https://jira.reactos.org/secure/attachment/48124/48124_CORE-14961-after-after.txt

**Comment by katahiromz** [comment]
AFTER x3:
https://jira.reactos.org/secure/attachment/48127/48127_CORE-14961-aaafter.txt
core-14961-aaafter

GO WITH IT!

File: dll/win32/shell32/CShellLink.cpp [hunk 1]
```diff
@@ -1721,6 +1721,9 @@ static HRESULT SHELL_PidlGetIconLocationW(PCIDLIST_ABSOLUTE pidl,
 HRESULT STDMETHODCALLTYPE CShellLink::GetIconLocation(UINT uFlags, PWSTR pszIconFile, UINT cchMax, int *piIndex, UINT *pwFlags)
 {
     HRESULT hr;
+
+    pszIconFile[0] = 0;
```
Comment (SergeGautherie) [inline dll/win32/shell32/CShellLink.cpp]: `L'\0'` value is more explicit.
Maybe use `*pszIconFile`, here and below?

Comment (learn-more) [reply dll/win32/shell32/CShellLink.cpp]: `pszIconFile[0]` is just fine.

Comment (katahiromz) [reply dll/win32/shell32/CShellLink.cpp]: OK.

File: dll/win32/shell32/CShellLink.cpp [hunk 2]
```diff
@@ -1888,9 +1897,10 @@ HRESULT STDMETHODCALLTYPE CShellLink::SetIconLocation(LPCWSTR pszIconPath, INT i
          */
         // FIXME: http://stackoverflow.com/questions/2976489/ishelllinkseticonlocation-translates-my-icon-path-into-program-files-which-i
         // if (PathFullyUnExpandEnvStringsW(pszIconPath, szUnExpIconPath, _countof(szUnExpIconPath)))
-        SHExpandEnvironmentStringsW(pszIconPath, szUnExpIconPath, _countof(szUnExpIconPath));
+        PathUnExpandEnvStringsW(pszIconPath, szUnExpIconPath, _countof(szUnExpIconPath));
         if (wcscmp(pszIconPath, szUnExpIconPath) != 0)
         {
+            pszIconPath = szUnExpIconPath;
```
Comment (HBelusca) [inline dll/win32/shell32/CShellLink.cpp]: Please move this just before the `pInfo = (LPEXP_SZ_LINK)...` line below.

Comment (katahiromz) [reply dll/win32/shell32/CShellLink.cpp]: OK.

**Comment by SergeGautherie** [comment]
For the record, with regard to CORE-14009, this PR seems to be more a side-effect workaround than a root fix.
Good improvement in the meantime anyway.

**Pull request merged**